File: 3. Behavioral Design Patterns/Template Method/C#/TemplateMethodExample.cs
// Template Method Pattern Example in C#
// DataProcessor with CSV and JSON implementations

using System;

namespace TemplateMethodExample
{
    abstract class DataProcessor
    {
        public void ProcessData()
        {
            LoadData();
            ProcessDataImpl();
            SaveData();
        }
        protected abstract void LoadData();
        protected abstract void ProcessDataImpl();
        protected abstract void SaveData();
    }

    class CSVProcessor : DataProcessor
    {
        protected override void LoadData() => Console.WriteLine("Loading CSV data");
        protected override void ProcessDataImpl() => Console.WriteLine("Processing CSV data");
        protected override void SaveData() => Console.WriteLine("Saving CSV data");
    }

    class JSONProcessor : DataProcessor
    {
        protected override void LoadData() => Console.WriteLine("Loading JSON data");
        protected override void ProcessDataImpl() => Console.WriteLine("Processing JSON data");
        protected override void SaveData() => Console.WriteLine("Saving JSON data");
    }

    class Program
    {
        static void Main()
        {
            DataProcessor csv = new CSVProcessor();
            csv.ProcessData();
            Console.WriteLine();
            DataProcessor json = new JSONProcessor();
            json.ProcessData();
        }
    }
}

File: 3. Behavioral Design Patterns/Template Method/Java/TemplateMethodExample.java
// Template Method Pattern Example in Java
// DataProcessor with CSV and JSON implementations

abstract class DataProcessor {
    public final void processData() {
        loadData();
        processDataImpl();
        saveData();
    }
    protected abstract void loadData();
    protected abstract void processDataImpl();
    protected abstract void saveData();
}

class CSVProcessor extends DataProcessor {
    protected void loadData() { System.out.println("Loading CSV data"); }
    protected void processDataImpl() { System.out.println("Processing CSV data"); }
    protected void saveData() { System.out.println("Saving CSV data"); }
}

class JSONProcessor extends DataProcessor {
    protected void loadData() { System.out.println("Loading JSON data"); }
    protected void processDataImpl() { System.out.println("Processing JSON data"); }
    protected void saveData() { System.out.println("Saving JSON data"); }
}

public class TemplateMethodExample {
    public static void main(String[] args) {
        DataProcessor csv = new CSVProcessor();
        csv.processData();
        System.out.println();
        DataProcessor json = new JSONProcessor();
        json.processData();
    }
}
